e2e/v3/jobs3/jobs3.go

```diff
@@ -0,0 +1,433 @@
+		select {
+		case <-ctx.Done():
+			must.Unreachable(sub.t, must.Sprint("timeout reached waiting for eval"))
+		default:
+		}
+
+		eval, _, err := evalAPI.Info(evalID, queryOpts)
+		must.NoError(sub.t, err)
+
+		sub.logf("checking eval: %s, status: %s", evalID, eval.Status)
+
+		switch eval.Status {
+
+		case nomadapi.EvalStatusComplete:
+			deploymentID = eval.DeploymentID
+			break EVAL
+		case nomadapi.EvalStatusFailed:
+			must.Unreachable(sub.t, must.Sprint("eval failed"))
+		case nomadapi.EvalStatusCancelled:
+			must.Unreachable(sub.t, must.Sprint("eval cancelled"))
+		default:
+			time.Sleep(1 * time.Second)
+		}
+
+		nextEvalID := eval.NextEval
+		if nextEvalID != "" {
+			evalID = nextEvalID
+			continue
+		}
+	}
+
+	switch *job.Type {
+	case "service":
+		// need to monitor the deployment until it is complete
+		depAPI := sub.nomadClient.Deployments()
+	DEPLOY:
+		for {
+
+			// check if we have passed timeout expiration
+			select {
+			case <-ctx.Done():
+				must.Unreachable(sub.t, must.Sprint("timeout reached waiting for deployment"))
+			default:
+			}
+
+			dep, _, err := depAPI.Info(deploymentID, queryOpts)
+			must.NoError(sub.t, err)
+
+			sub.logf("checking deployment: %s, status: %s", dep.ID, dep.Status)
+
+			switch dep.Status {
+			case nomadapi.DeploymentStatusBlocked:
+				must.Unreachable(sub.t, must.Sprint("deployment is blocked"))
+			case nomadapi.DeploymentStatusCancelled:
+				must.Unreachable(sub.t, must.Sprint("deployment is cancelled"))
+			case nomadapi.DeploymentStatusFailed:
+				must.Unreachable(sub.t, must.Sprint("deployment is failed"))
+			case nomadapi.DeploymentStatusPaused:
+				must.Unreachable(sub.t, must.Sprint("deployment is paused"))
+			case nomadapi.DeploymentStatusPending:
+				break
+			case nomadapi.DeploymentStatusRunning:
+				break
+			case nomadapi.DeploymentStatusSuccessful:
+				sub.logf("deployment %s was a success", dep.ID)
+				break DEPLOY
+			case nomadapi.DeploymentStatusUnblocking:
+				must.Unreachable(sub.t, must.Sprint("deployment is unblocking"))
+			default:
+				break
+			}
+			time.Sleep(1 * time.Second)
+		}
+	// todo: more job types
+	default:
+	}
+
+}
+
+func (sub *Submission) waitAlloc(group, id string) {
+	queryOpts := sub.queryOptions()
+	allocAPI := sub.nomadClient.Allocations()
+
+	// todo: respect timeout
+
+ALLOCATION:
+	for {
+		latest, _, err := allocAPI.Info(id, queryOpts)
+		must.NoError(sub.t, err)
+
+		status := latest.ClientStatus
+		sub.logf("wait for %q allocation %s, status: %s", group, id, status)
+		switch status {
+		case nomadapi.AllocClientStatusLost:
+			must.Unreachable(sub.t, must.Sprintf("group %q allocation %s lost", group, id))
+		case nomadapi.AllocClientStatusFailed:
+			must.Unreachable(sub.t, must.Sprintf("group %q allocation %s failed", group, id))
+		case nomadapi.AllocClientStatusPending:
+			break
+		case nomadapi.AllocClientStatusRunning:
+			break
+		case nomadapi.AllocClientStatusComplete:
+			break ALLOCATION
```

Comment:
There's also `unknown` which seem to be missing from `api` 😬

Reply:
yikes! filed https://github.com/hashicorp/nomad/issues/17688